hazelcast/src/main/java/com/hazelcast/kubernetes/KubernetesClient.java

```diff
@@ -608,4 +688,88 @@ public String toString() {
             return String.format("%s:%s", ip, port);
         }
     }
+
+    final class StsMonitor implements Runnable {
+        private String latestResourceVersion;
+        private RuntimeContext latestRuntimeContext;
+
+        /**
+         * Initializes and watches information about the StatefulSet in which Hazelcast is being executed.
+         * See <a href="https://kubernetes.io/docs/reference/using-api/api-concepts/#efficient-detection-of-changes">
+         * Efficient detection of changes on Kubernetes API reference</a>.
+         *
+         * Important: If this thread starves, then timely updates may be stalled and shutdown hook
+         * may not act on the latest cluster information.
+         */
+        @Override
+        public void run() {
+            String stsUrlString = String.format("%s/apis/apps/v1/namespaces/%s/statefulsets", kubernetesMaster,
```

Comment:
maybe no need to run this if it is not deployed as stateless app?

Reply:
I am not sure if or how an application that runs within a pod can figure out if it is deployed as part of statefulset or not.

The condition for the `StsMonitor` thread to be started is persistence to be enabled + kubernetes discovery being in use. It wouldn't make much sense to run with persistence enabled outside of a statefulset, so we are using this config as a proxy for "am I running in a stateful set?".